packages/compiler/core/parser.ts

```diff
@@ -1262,6 +1296,152 @@ export function parse(code: string | SourceFile, options: ParseOptions = {}): Ca
+  }
+
+  function parseModifier(): Modifier | undefined {
+    switch (token()) {
+      case Token.ExternKeyword:
+        return parseExternKeyword();
+      default:
+        return undefined;
+    }
+  }
+
+  function parseDecoratorDeclarationStatement(
+    pos: number,
+    modifiers: Modifier[]
+  ): DecoratorDeclarationStatementNode {
+    const modifierFlags = modifiersToFlags(modifiers);
+    parseExpected(Token.DecKeyword);
+    const id = parseIdentifier();
+    let [target, ...parameters] = parseFunctionParameters();
+    if (target === undefined) {
+      error({ code: "decorator-decl-target", target: { pos, end: previousTokenEnd } });
+      target = {
+        kind: SyntaxKind.FunctionParameter,
+        id: createMissingIdentifier(),
+        value: createMissingIdentifier(),
+        optional: false,
+        rest: false,
+        ...finishNode(pos),
+      };
+    }
+    if (target.optional) {
+      error({ code: "decorator-decl-target", messageId: "required" });
+    }
+    parseExpected(Token.Semicolon);
+    return {
+      kind: SyntaxKind.DecoratorDeclarationStatement,
+      modifiers,
+      modifierFlags,
+      id,
+      target,
+      parameters,
+      ...finishNode(pos),
+    };
+  }
+
+  function parseFunctionDeclarationStatement(
+    pos: number,
+    modifiers: Modifier[]
+  ): FunctionDeclarationStatementNode {
+    const modifierFlags = modifiersToFlags(modifiers);
+    parseExpected(Token.FnKeyword);
+    const id = parseIdentifier();
+    const parameters = parseFunctionParameters();
+    parseExpected(Token.Colon);
+    const returnType = parseExpression();
+    parseExpected(Token.Semicolon);
+    return {
+      kind: SyntaxKind.FunctionDeclarationStatement,
+      modifiers,
+      modifierFlags,
+      id,
+      parameters,
+      returnType,
+      ...finishNode(pos),
+    };
+  }
+
+  function parseFunctionParameters(): FunctionParameterNode[] {
+    const parameters = parseList<typeof ListKind.FunctionParameters, FunctionParameterNode>(
+      ListKind.FunctionParameters,
+      parseFunctionParameter
+    );
+
+    let foundOptional = false;
+    for (const [index, item] of parameters.entries()) {
+      if (!item.optional && foundOptional) {
+        error({ code: "required-parameter-first", target: item });
+        continue;
+      }
+
+      if (item.optional) {
+        foundOptional = true;
+      }
+
+      if (item.rest && item.optional) {
+        error({ code: "rest-parameter-required", target: item });
+      }
+      if (item.rest && index !== parameters.length - 1) {
+        error({ code: "rest-parameter-last", target: item });
+      }
+    }
+    return parameters;
+  }
+
+  function parseFunctionParameter(): FunctionParameterNode {
+    const pos = tokenPos();
+    const rest = parseOptional(Token.Ellipsis);
+    const id = parseIdentifier("property");
+
+    const optional = parseOptional(Token.Question);
+    parseExpected(Token.Colon);
+    const value = parseExpression();
```

Comment:
`type` feels like a better name than `value`, to me? Unless I am misunderstanding what this is.

Reply:
no it is the type, I think I kept the pattern used for model properties but yeah `type` make more sense